ykrt/src/trace/hwt/mapper.rs

```diff
@@ -1,37 +1,119 @@
 //! The mapper translates a hwtracer trace into an IR trace.
 
 use crate::trace::{AOTTraceIterator, AOTTraceIteratorError, InvalidTraceError, TraceAction};
-use hwtracer::llvm_blockmap::LLVM_BLOCK_MAP;
-use hwtracer::Trace;
-use std::error::Error;
+use hwtracer::{llvm_blockmap::LLVM_BLOCK_MAP, Block, HWTracerError, Trace};
 use ykaddr::{
     addr::{vaddr_to_obj_and_off, vaddr_to_sym_and_obj},
     obj::SELF_BIN_PATH,
 };
 
+/// Map the *machine* blocks of the specified trace into LLVM IR blocks.
+///
+/// Each entry in the returned trace is either a "mapped block" identifying a successfully
+/// mapped LLVM IR block, or an unsuccessfully mapped "unmappable block" (an unknown region of
+/// code spanning at least one machine block).
 pub(crate) struct HWTTraceIterator {
-    trace: std::vec::IntoIter<TraceAction>,
+    hwt_iter: Box<dyn Iterator<Item = Result<Block, HWTracerError>> + Send>,
+    /// The next [TraceAction]`s we will produce when `next` is called. We need this intermediary
+    /// to allow us to deduplicate mapped/unmapped blocks. This will be empty on the first
+    /// iteration and from then on will always have at least one [TraceAction] in it.
+    upcoming: Vec<TraceAction>,
+    /// How many [TraceAction]s have been generated so far? We use this to know if the underlying
+    /// trace is too long.
+    tas_generated: usize,
 }
 
 impl AOTTraceIterator for HWTTraceIterator {}
 
 impl Iterator for HWTTraceIterator {
     type Item = Result<TraceAction, AOTTraceIteratorError>;
     fn next(&mut self) -> Option<Self::Item> {
-        self.trace.next().map(|x| Ok(x))
+        while self.upcoming.len() < 2 {
```

Comment:
Why 2?

Reply:
To allow deduplication: you always need to have 1 element in this buffer in order to know if the next is a duplicate.

Reply:
Fixed in 86f0eb7.